### PR TITLE
Implement gcd, lcm

### LINKED ISSUE #40018: Implementing gcd, lcm
From #38349.

cc @mruberry

### PR DESCRIPTION
**Pull request by aayn**
Resolves #40018.

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -2496,6 +2496,30 @@ def merge_dicts(*dicts):
             [ 4,  3]])
 """)
 
+
+add_docstr(torch.gcd,
```
Comment (aayn) [inline torch/_torch_docs.py:2500]: Still need to finish this.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -238,6 +238,8 @@ def method_tests():
         ('erfc', uniform_scalar(requires_grad=True), NO_ARGS, 'scalar', (True,)),
         ('erfinv', torch.rand(S, S, S).clamp(-0.9, 0.9), NO_ARGS),
         ('erfinv', normal_scalar_clamp(-0.9, 0.9, requires_grad=True), NO_ARGS, 'scalar'),
+        ('gcd', (S, S, S), ((S, S, S),), '', (False,)),
```
Comment (aayn) [inline torch/testing/_internal/common_methods_invocations.py:241]: How do I change this so that `gcd` gets invoked only with integral arguments?

@mruberry

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:241]: You don't need to add gcd or lcm to common methods invocations, since it's mostly intended for testing gradients, which you're not dealing with.

Comment (aayn) [reply torch/testing/_internal/common_methods_invocations.py:241]: Okay. In that case, I'm ready for a review.

File: aten/src/ATen/native/Math.h [hunk 3]
```diff
@@ -267,3 +267,15 @@ static inline float calc_digamma(float x) {
...
+static inline scalar_t calc_gcd(scalar_t a, scalar_t b) {
+  a = std::abs(a);
+  b = std::abs(b);
+  while (a != 0) {
+    scalar_t c = a;
+    a = b % a;
+    b = c; 
+  }
+  return b;
+}
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:281]: Newline

File: aten/src/ATen/native/Math.h [hunk 4]
```diff
@@ -267,3 +267,15 @@ static inline float calc_digamma(float x) {
 }
 
 inline c10::BFloat16 calc_erfinv(c10::BFloat16 a) { return calc_erfinv(float(a)); }
+
+template <typename scalar_t>
+static inline scalar_t calc_gcd(scalar_t a, scalar_t b) {
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:272]: What does this return if a is zero? If b is zero?

Comment (mruberry) [reply aten/src/ATen/native/Math.h:272]: (We should just add a comment explaining the behavior, since we can't return NaN for gcd(0, 0) here)

I suppose we could return an optional, but that seems like a lot.

Comment (mruberry) [reply aten/src/ATen/native/Math.h:272]: Also, better annotate this with C10_HOST_DEVICE:
 https://github.com/pytorch/pytorch/blob/a2ef54c598261ad9fd15e1e85c71f15055652572/c10/macros/Macros.h#L157 

since you're calling it from CUDA and C++. 

And you can use the same type traits trick as `calc_erfinv` (with `is_integral` instead of `is_floating_point`):

https://github.com/pytorch/pytorch/blob/40228ef2e59dced45b1b39819e57ea150e14b4ce/aten/src/ATen/native/Math.h#L66

File: aten/src/ATen/native/cuda/Math.cuh [hunk 5]
```diff
@@ -88,5 +88,19 @@ static inline __host__ __device__ scalar_t calc_trigamma(scalar_t in) {
   return static_cast<scalar_t>(sign * result);
 }
 
+template <typename scalar_t>
+static inline __host__ __device__ scalar_t calc_gcd(scalar_t a, scalar_t b) {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Math.cuh:92]: You can use: 

https://github.com/pytorch/pytorch/blob/a2ef54c598261ad9fd15e1e85c71f15055652572/c10/macros/Macros.h#L157

File: aten/src/ATen/native/cuda/Math.cuh [hunk 6]
```diff
@@ -88,5 +88,19 @@ static inline __host__ __device__ scalar_t calc_trigamma(scalar_t in) {
...
+  using accscalar_t = at::acc_type<scalar_t, /*is_cuda=*/true>;
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Math.cuh:93]: Is the accumulate type ever different for the integer types?

File: docs/source/name_inference.rst [hunk 7]
```diff
@@ -109,6 +109,8 @@ If you don't see an operation listed here, but it would help your use case, plea
    :meth:`Tensor.floor_`,None
    ":meth:`Tensor.frac`, :func:`torch.frac`",:ref:`keeps_input_names-doc`
    :meth:`Tensor.frac_`,None
+   ":meth:`Tensor.gcd`, :func:`torch.gcd`",:ref:`keeps_input_names-doc`
```
Comment (mruberry) [inline docs/source/name_inference.rst:112]: Let's remove this for now. Named tensor support is on the back burner and these functions don't directly support named tensors since they're implementing a new kernel.

File: test/test_torch.py [hunk 8]
```diff
@@ -19182,6 +19182,12 @@ def inner(self, device, dtype):
     ('expand_as', '', _new_t((_M, 1, _M)), lambda t, d: [_new_t((_M, 4, _M))(t, d)],
         1e-5, 1e-5, 1e-5, _types, _cpu_types, False),
     ('fill_', '', _medium_2d, lambda t, d: [_number(3.14, 3, t)], 1e-3, 1e-5, 1e-5, _types, _cpu_types, False),
+    ('gcd', '', _small_3d, lambda t, d: [_small_3d(t, d)], 1e-1, 1e-0, 1e-5,
```
Comment (mruberry) [inline test/test_torch.py:19185]: For the tolerances here, can they not all be 0?

File: torch/_tensor_docs.py [hunk 9]
```diff
@@ -1320,6 +1320,20 @@ def add_docstr_all(method, docstr):
 See :func:`torch.gather`
 """)
 
+add_docstr_all('gcd',
+               r"""
+gcd(value) -> Tensor
```
Comment (mruberry) [inline torch/_tensor_docs.py:1325]: "value"->"other" to be consistent with the documentation below?

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -2496,6 +2496,31 @@ def merge_dicts(*dicts):
...
+           r"""
+gcd(input, other, out=None) -> LongTensor
```
Comment (mruberry) [inline torch/_torch_docs.py:2502]: "-> Tensor", not "LongTensor"

File: torch/_torch_docs.py [hunk 11]
```diff
@@ -2496,6 +2496,31 @@ def merge_dicts(*dicts):
...
+
+Computes the element-wise greatest common divisor (GCD).
+
+The two tensors involved in the operation must be integral.
+
+Args:
+    input (LongTensor)
```
Comment (mruberry) [inline torch/_torch_docs.py:2509]: "LongTensor" -> "Tensor"

File: torch/_torch_docs.py [hunk 12]
```diff
@@ -2496,6 +2496,31 @@ def merge_dicts(*dicts):
...
+    other (LongTensor)
+    {out}
```
Comment (mruberry) [inline torch/_torch_docs.py:2511]: out is a kwarg, see how add treats it

File: torch/_torch_docs.py [hunk 13]
```diff
@@ -2496,6 +2496,31 @@ def merge_dicts(*dicts):
             [ 4,  3]])
 """)
 
+
+add_docstr(torch.gcd,
+           r"""
+gcd(input, other, out=None) -> LongTensor
+
+Computes the element-wise greatest common divisor (GCD).
```
Comment (mruberry) [inline torch/_torch_docs.py:2504]: "... of input and other."

Comment (mruberry) [reply torch/_torch_docs.py:2504]: Let's be pedantic and be clear the function defines gcd(0, 0)=0.

File: torch/_torch_docs.py [hunk 14]
```diff
@@ -2496,6 +2496,31 @@ def merge_dicts(*dicts):
...
+
+The two tensors involved in the operation must be integral.
```
Comment (mruberry) [inline torch/_torch_docs.py:2506]: "Both input and other must have integer dtypes."

File: torch/_torch_docs.py [hunk 15]
```diff
@@ -2966,6 +2991,30 @@ def merge_dicts(*dicts):
     torch.return_types.kthvalue(values=tensor([[4., 5., 6.]]), indices=tensor([[1, 1, 1]]))
 """.format(**single_dim_common))
 
+add_docstr(torch.lcm,
+           r"""
+lcm(input, other, out=None) -> LongTensor
+
+Computes the element-wise least common multiple (LCM).
```
Comment (mruberry) [inline torch/_torch_docs.py:2998]: See above notes.

**Review by mruberry — CHANGES_REQUESTED** [review]
This is an excellent and well-written PR implementing two interesting functions. I've requested a few small changes for clarity.

File: torch/_torch_docs.py (hunk 15 above)
Comment (mruberry) [reply torch/_torch_docs.py:2998]: In particular, what is lcm(a, 0)?

Should there be a test in test_torch.py explicitly for the gcd and lcm edge cases where one of the two inputs?

File: aten/src/ATen/native/Math.h (hunk 4 above)
Comment (aayn) [reply aten/src/ATen/native/Math.h:272]: If `a` is 0, we return `b` (this is the case even when `b` is also 0). If `b` is 0 and `a` is nonzero then the while loop is executed once, `b` takes the value of `a` and is returned. In short, this will never return NaN, with `gcd(a, 0) = a`; `gcd(0, b) = b`; `gcd(0, 0) = 0`.

File: aten/src/ATen/native/cuda/Math.cuh (hunk 6 above)
Comment (aayn) [reply aten/src/ATen/native/cuda/Math.cuh:93]: I thought `accscalar_t` stood for "accelerated scalar type" (as in accelerated with CUDA), my bad. I'll change that.

File: test/test_torch.py (hunk 8 above)
Comment (aayn) [reply test/test_torch.py:19185]: Yes, you're right.

File: torch/_torch_docs.py (hunk 15 above)
Comment (aayn) [reply torch/_torch_docs.py:2998]: Yeah, I think having explict tests is a good idea (I tested these edge cases against the numpy implementation manually). And, lcm(a, 0) = 0 (which is also the case in the numpy implementation).

File: aten/src/ATen/native/cuda/Math.cuh (hunk 5 above)
Comment (aayn) [reply aten/src/ATen/native/cuda/Math.cuh:92]: All other functions in that header file also don't use that macro. Should I change them too?

File: torch/_torch_docs.py (hunk 15 above)
Comment (aayn) [reply torch/_torch_docs.py:2998]: On second thought, doesn't [this line](https://github.com/pytorch/pytorch/blob/a79b41684742de3133b918b85f08b43f147098b3/test/test_torch.py#L19482) already account for this?

**Comment by aayn** [comment]
Thanks! Do you know why the XLA tests are failing?

**Comment by JackCaoG** [comment]
Looking at the log `Cannot access data pointer of Tensor that doesn't have storage` It seems like test is trying to access the storage while xla does not have one. I will submit a pr to disable those tests.

File: aten/src/ATen/native/Math.h (hunk 4 above)
Comment (mruberry) [reply aten/src/ATen/native/Math.h:272]: Good. I just want to be clear about the (0, 0) behavior.

File: aten/src/ATen/native/cuda/Math.cuh (hunk 5 above)
Comment (mruberry) [reply aten/src/ATen/native/cuda/Math.cuh:92]: No. Old code that's working is best left undisturbed. Thank you for asking, though.

File: torch/_torch_docs.py (hunk 15 above)
Comment (mruberry) [reply torch/_torch_docs.py:2998]: You would think so because our test suite is so confusing, but I think that function is for the Math tests and your tests are in the method test section. I'm actually about to overhaul all of this so it's not super important where you test the extremal values. You can probably just add lcm to the math test portion, but some people have struggled to fit their functions in there. You would do that by adding to `torch_op_tests` on line 19546.

**Comment by mruberry** [comment]
> Looking at the log `Cannot access data pointer of Tensor that doesn't have storage` It seems like test is trying to access the storage while xla does not have one. I will submit a pr to disable those tests.

You can't disable them because they're not even in master yet ;)

It's not super surprising these new tests are failing on XLA. You can disable the test just on XLA by adding the `@onlyOnCPUAndCUDA` decorator to your tests. For the method tests you can see how decorators are applied to them in the test for `qr`, for example, where the `skipCUDAIfNoMagma` decorator is applied.

**Comment by aayn** [comment]
I've updated the PR: adding edge-case tests for gcd and lcm, and reflected other suggestions.

**Comment by aayn** [comment]
What's with these build errors?

**Comment by mruberry** [comment]
> What's with these build errors? Should I rebase and push again?

Looks like master was in a bad state earlier today. Sorry about that @aayn. If you merge with the latest master and push they should be OK.

**Comment by aayn** [comment]
Checks have passed! @mruberry

**Review by mruberry — APPROVED** [review]
Awesome! Nice work, @aayn. Let me know if you need another project ;)

**Comment by mruberry** [comment]
Sorry @aayn, looks like we got a small merge conflict at the last moment. Would you please resolve it and I'll restart the land process?

**Comment by aayn** [comment]
@mruberry Done. And yeah, I'd love work on another project.

**Comment by mruberry** [comment]
> @mruberry Done. And yeah, I'd love to work on another project.

Great! This should land soon. 

If you'd like to implement a similar function with a twist there's [divmod](https://numpy.org/doc/1.18/reference/generated/numpy.divmod.html?highlight=divmod#numpy.divmod), something a little different is [vdot](https://numpy.org/doc/1.18/reference/generated/numpy.vdot.html?highlight=vdot#numpy.vdot), and something completely different is [broadcast_to](https://numpy.org/doc/1.18/reference/generated/numpy.broadcast_to.html?highlight=broadcast_to#numpy.broadcast_to). Would any of those be interesting?
